### PR TITLE
Fix HPOS unit tests for WC Order unit test classes

### LINKED ISSUE #36685: [HPOS] Add compatibility for WC Order unit test classes.
This issue scopes making tests in `WC_Tests_WC_Order_Query`, `WC_Abstract_Order_Test` and `WC_Order_Data_Store_CPT_Test` classes compatible with HPOS by fixing test or core code, as needed.

Please fork from #36650 for workflow to run tests in HPOS context.

Towards #36376

### PR DESCRIPTION
**Pull request by vedanshujain**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #36685 
Targets #36832

We have two changes in this PR:
#### 1. Fetching order first when getting prop.

In the post-type store, we used the `get_metadata` function to get the value of any particular prop. In OrderTableDataStore, we were earlier returning values that already existed in the Order object when getting a prop, (for example recorded_sales). Unfortunately, this behavior is different in OrderTableDataStore and Post store, since post store used to do a `get_metadata` call, which in many scenarios will return more refreshed data than what is stored in the order object, which could be quite stale.

In this PR, we make this consistent by fetching the order object again before returning the prop. This will load the order either from the cache when there are no updates to the order object, or it will fetch from the DB when the order is being written somewhere and the cache is validated.

#### 2. Map legacy date props to proper date props with correct timezone.

In OrderTableQuery, we were incorrectly mapping several legacy args, such as `post_date` to new prop such as `date_created_gmt`. The error here is that timezones aren't accounted for, and a localized date param is being mapped to a GMT param. This PR fixes this issue by correct mapping the old params to new and taking account of the timezone in the process. We also account for these timezones in the `process_date_args` method in OrdersTableQuery.

- [x] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

Unit test coverage is sufficient to test these backend changes, make sure that tests for `WC_Tests_WC_Order_Query`, `WC_Abstract_Order_Test` and `WC_Order_Data_Store_CPT_Test` passes for both orders and CPT.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by vedanshujain** [comment]
Note that the test `Run unit tests on PR / PHP 7.4 WP latest HPOS ` will fail and is expected. Verify that none of the failures belong to WC_Tests_WC_Order_Query, WC_Abstract_Order_Test and WC_Order_Data_Store_CPT_Test. The Highlight templates changes seems like a build error, which I will fix in #36832

**Comment by vedanshujain** [comment]
Note that I am checking `Highlight templates changes` test in #36685 but it looks like a build error rather than something in the code. Test `Run unit tests on PR / PHP 7.4 WP latest HPOS` is expected to fail, as will pass when all PRs in the stack is merged, however there should be no failing test for WC_Tests_WC_Order_Query, WC_Abstract_Order_Test and WC_Order_Data_Store_CPT_Test .

File: plugins/woocommerce/tests/php/includes/class-wc-order-factory-test.php [hunk 1]
```diff
@@ -5,6 +5,35 @@
...
+
+	/**
+	 * Disable COT before the test.
+	 *
+	 * @return void
+	 */
+	public function setUp(): void {
+		parent::setUp();
+		$this->cot_state = \Automattic\WooCommerce\Utilities\OrderUtil::custom_orders_table_usage_is_enabled();
+		\Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::toggle_cot( false );
```
Comment (Konamiman) [inline plugins/woocommerce/tests/php/includes/class-wc-order-factory-test.php]: I think these (and similar others in the same file) should be simplified into `use` statements for the namespaces, for readability and for consistency with what we normally do.

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php [hunk 2]
```diff
@@ -1855,7 +1868,8 @@ private function assert_get_prop_via_ds_object_and_metadata( array $props, WC_Or
 	 * @testDox Legacy getters and setters for props migrated from data stores should be set/reset properly.
 	 */
 	public function test_legacy_getters_setters() {
-		$order_id = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_complex_wp_post_order();
+		$this->toggle_cot( true );
+		$order_id = \Automattic\WooCommerce\RestApi\UnitTests\Helpers\OrderHelper::create_complex_data_store_order( $this->sut );
```
Comment (Konamiman) [inline plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: Same: please change to `use`s and simple class names.

**Review by Konamiman — CHANGES_REQUESTED** [review]
I have requested a couple of minor styling changes, otherwise it looks good. There are a few unit tests that are failing in CI but they all pass for me locally.

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php (hunk 2 above)
Comment (vedanshujain) [reply plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: Done in https://github.com/woocommerce/woocommerce/pull/36843/commits/fb356b686e10f779fb7e6a50908eefe9b7dea045

File: plugins/woocommerce/tests/php/includes/class-wc-order-factory-test.php (hunk 1 above)
Comment (vedanshujain) [reply plugins/woocommerce/tests/php/includes/class-wc-order-factory-test.php]: Done in https://github.com/woocommerce/woocommerce/pull/36843/commits/fb356b686e10f779fb7e6a50908eefe9b7dea045

**Comment by vedanshujain** [comment]
> There are a few unit tests that are failing in CI but they all pass for me locally.

This is expected, and they will start to pass as I continue to fold this PR stack.

**Comment by vedanshujain** [comment]
Note that this was deleted and remerged in #37130

**Pull request merged**